### PR TITLE
Fix plugin icon URLs in electron & Fix custom icon themes

### PR DESCRIPTION
**Pull request by owlJaeger**
Signed-off-by: Luca Jaeger <owl.jaeger@gmail.com>

#### What it does
Fixes Issues:
- #6834 
- #7040 

#6834 Fixed by:
- [plugin-protocol.ts](https://github.com/eclipse-theia/theia/pull/7583/files#diff-2a5766562161792548814f3e490b89ff):
  - Replace `PluginModel.packagePath` with a backend-created `FileUri` passed on as `PluginModel.packageUri` instead of `PluginModel.packagePath`
- [scanner-vscode.ts](https://github.com/eclipse-theia/theia/pull/7583/files#diff-2894f6e74c355c6d2c8a958123199bc4) & [scanner-theia.ts](https://github.com/eclipse-theia/theia/pull/7583/files#diff-eff368badd61fda201e6cbcf60009afe):
  - Pass on a backend-generated `FileUri` through the `PluginModel.packageUri` property.
- [plugin-api-rpc.ts](https://github.com/eclipse-theia/theia/pull/7583/files#diff-7eea09f2b0c581de589cb87e1127fb80):
  - Updated the empty Model to include `packageUri`
- [plugin-icon-theme-service.ts](https://github.com/eclipse-theia/theia/compare/master...owlJaeger:master#diff-fd906c4e43a8828ebde6f8d5fa9cd22f):
  - Updated to use `packageUri` instead of `packagePath` and fix the `relativePath` being `undefined` on Windows.
  - Using `Endpoint`s to fix it for electron specifically.

#7040 Fixed by:
- Using `Endpoint`s in these files:
  - [plugin-shared-style.ts](https://github.com/eclipse-theia/theia/compare/master...owlJaeger:master#diff-fa8282987c4c581621df7483622e5f95)
  - [plugin-view-registry.ts](https://github.com/eclipse-theia/theia/compare/master...owlJaeger:master#diff-b643d2a05314e4c09a9d102ca71aced3)

#### How to test
- #6834: Test by building the `electron` and `browser` [example apps](https://github.com/eclipse-theia/theia/tree/master/examples) in the [theia](https://github.com/eclipse-theia/theia) repository and using the Seti File Icon Theme or download other icon themes through `vsx-registry`.
- #7040: Test by building the `electron` [example app](https://github.com/eclipse-theia/theia/tree/master/examples/electron) in the [theia](https://github.com/eclipse-theia/theia) repository and install an extension like `Gitlens`.

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/plugin-ext/src/main/browser/view/plugin-view-registry.ts [hunk 1]
```diff
@@ -14,32 +14,32 @@
...
-import { PluginSharedStyle } from '../plugin-shared-style';
-import { DebugWidget } from '@theia/debug/lib/browser/view/debug-widget';
-import { PluginViewWidget, PluginViewWidgetIdentifier } from './plugin-view-widget';
-import { SCM_VIEW_CONTAINER_ID, ScmContribution } from '@theia/scm/lib/browser/scm-contribution';
-import { EXPLORER_VIEW_CONTAINER_ID } from '@theia/navigator/lib/browser';
-import { FileNavigatorContribution } from '@theia/navigator/lib/browser/navigator-contribution';
-import { DebugFrontendApplicationContribution } from '@theia/debug/lib/browser/debug-frontend-application-contribution';
-import { Disposable, DisposableCollection } from '@theia/core/lib/common/disposable';
-import { CommandRegistry } from '@theia/core/lib/common/command';
-import { MenuModelRegistry } from '@theia/core/lib/common/menu';
```
Comment (vince-fugnitto) [inline packages/plugin-ext/src/main/browser/view/plugin-view-registry.ts:33]: @owlJaeger can you revert these changes, they are unnecessary and are noisy during reviews.

Comment (owlJaeger) [reply packages/plugin-ext/src/main/browser/view/plugin-view-registry.ts:33]: Sure, will take care of it.

**Comment by vince-fugnitto** [comment]
Please be sure to properly sign the [Eclipse ECA](https://github.com/eclipse-theia/theia/blob/master/CONTRIBUTING.md#eclipse-contributor-agreement), I see that your sign-off email is not yet registered:

<div align='center'>

Screen Shot 2020-04-15 at 12 40 06 PM


</div>

File: packages/plugin-ext/src/main/browser/view/plugin-view-registry.ts (hunk 1 above)
Comment (owlJaeger) [reply packages/plugin-ext/src/main/browser/view/plugin-view-registry.ts]: Done

Comment (owlJaeger) [reply packages/plugin-ext/src/main/browser/view/plugin-view-registry.ts]: Done

**Comment by owlJaeger** [comment]
Should be registered now

File: packages/core/src/common/uri.ts [hunk 2]
```diff
@@ -213,4 +214,11 @@ export default class URI {
         return result;
     }
 
+    static file(path: string): URI {
+        if (isWindows) {
+            return new URI(Uri.file(path.replace(/\\/g, '/')));
+        } else {
+            return new URI(Uri.file(path));
+        }
+    }
```
Comment (vince-fugnitto) [inline packages/core/src/common/uri.ts:223]: This is the part I was unsure about, @akosyakov do you have any comments about it?

Comment (akosyakov) [reply packages/core/src/common/uri.ts:223]: it should be removed. There is `FileUri` and it can be used only on the backend. Please see URIs/paths guidelines: https://github.com/eclipse-theia/theia/wiki/Coding-Guidelines#uripath

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,12 @@
 # Change Log
 
+## v1.1.1
```
Comment (akosyakov) [inline CHANGELOG.md:3]: 1.1.0 is next release

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,5 +1,12 @@
...
+
+- [core] added static function to fix Windows filepaths that are using `\` instead of `/`
+- [plugin-ext] fixed custom Icon Themes not working
```
Comment (akosyakov) [inline CHANGELOG.md:6]: there is one item for each PR with a reference and sub points if needed, please see records below for an example

File: packages/plugin-ext/src/main/browser/plugin-icon-theme-service.ts [hunk 5]
```diff
@@ -127,7 +128,7 @@ export class PluginIconTheme extends PluginIconThemeDefinition implements IconTh
     @postConstruct()
     protected init(): void {
         Object.assign(this, this.definition);
-        this.packageUri = new URI(this.packagePath);
+        this.packageUri = URI.file(this.packagePath);
```
Comment (akosyakov) [inline packages/plugin-ext/src/main/browser/plugin-icon-theme-service.ts:131]: It cannot be used here: https://github.com/eclipse-theia/theia/wiki/Coding-Guidelines#frontend-fs-path

File: packages/core/src/common/uri.ts (hunk 2 above)
Comment (owlJaeger) [reply packages/core/src/common/uri.ts:223]: Will take care

File: packages/plugin-ext/src/main/browser/plugin-icon-theme-service.ts (hunk 5 above)
Comment (owlJaeger) [reply packages/plugin-ext/src/main/browser/plugin-icon-theme-service.ts:131]: It wouldn't for me because the `:` and `\` would mess up the Uri and I couldn't get `relative` to work... I'll try again with path's and discard the `static file` function entirely.

File: CHANGELOG.md (hunk 3 above)
Comment (owlJaeger) [reply CHANGELOG.md:3]: Oh, I see. Will also take are of that.

File: packages/plugin-ext/src/main/browser/plugin-icon-theme-service.ts (hunk 5 above)
Comment (owlJaeger) [reply packages/plugin-ext/src/main/browser/plugin-icon-theme-service.ts:131]: So the packageUri looks like: `C:\...\...` while the normalized icon uri path looks like: `/c:/.../...`.
To use the relative function the packageUri path also has to be in the form: `/c:/.../...` with only forwardslashes and one forwardslash at the beginning. But by constructing a `URI` object with the packagePath the path will result in: `/User/...`. I could change the [normalizeDrive](https://github.com/eclipse-theia/theia/blob/master/packages/core/src/common/path.ts) function to automatically insert the missing `/` at the beginning. But then I'd have to use the Path class inside of [plugin-icon-theme-service.ts](https://github.com/eclipse-theia/theia/blob/master/packages/plugin-ext/src/main/browser/plugin-icon-theme-service.ts) or I could change the URI constructor in [uri.ts](https://github.com/eclipse-theia/theia/blob/master/packages/core/src/common/uri.ts) to use the string ( if one is given )  and first construct a Path that then is automatically corrected with the beginning `/` and use the resulting string to parse with the `Uri.parse` function. I'll update my repo and then you can decide to do that or to just do it in the [plugin-icon-theme-service.ts](https://github.com/eclipse-theia/theia/blob/master/packages/plugin-ext/src/main/browser/plugin-icon-theme-service.ts) and use `Path` there.

Comment (akosyakov) [reply packages/plugin-ext/src/main/browser/plugin-icon-theme-service.ts]: We should avoid changing core types as URI and Path.

Comment (owlJaeger) [reply packages/plugin-ext/src/main/browser/plugin-icon-theme-service.ts]: For me it doesn't even work in the browser. The packagePath looks like: `C:\...\...` for me and after the new URI was created the normalized path of that Uri will result in: `\Users\luca\Desktop\tmp\theia\plugins\vscode-builtin-icon-theme-seti\extension\` while the normalized path of the icon uri is: `/c:/Users/luca/Desktop/tmp/theia/plugins/vscode-builtin-icon-theme-seti/extension/icons/seti.woff`. To get the relative path both must start the same way though: 
```
    relative(path: Path): Path | undefined {
        if (this.raw === path.raw) {
            return new Path('');
        }
        if (!this.raw || !path.raw) {
            return undefined;
        }
        const raw = this.base ? this.raw + Path.separator : this.raw;
==>    if (!path.raw.startsWith(raw)) {
            return undefined;
        }
        const relativePath = path.raw.substr(raw.length);
        return new Path(relativePath);
    }
```
I agree that I shouldn't modify URI and just use Path inside of [plugin-icon-theme-service.ts](https://github.com/eclipse-theia/theia/blob/master/packages/plugin-ext/src/main/browser/plugin-icon-theme-service.ts). But I also think the change in [path.ts](https://github.com/eclipse-theia/theia/blob/master/packages/core/src/common/path.ts) is good, because the normalization should make them equal... A normalized `C:/...` Path should equal a normalized `/c:/...` Path. My new changes dispense the change of [uri.ts](https://github.com/eclipse-theia/theia/blob/master/packages/core/src/common/uri.ts) but still include the `Path` fix. It is easy to fix it inside of [path.ts](https://github.com/eclipse-theia/theia/blob/master/packages/core/src/common/path.ts) without having to add long checks because they're already there in `Path.ts`.

Comment (akosyakov) [reply packages/plugin-ext/src/main/browser/plugin-icon-theme-service.ts]: What is the value of `packagePath` for you?

Comment (akosyakov) [reply packages/plugin-ext/src/main/browser/plugin-icon-theme-service.ts]: The troubles come from the fact that we violate https://github.com/eclipse-theia/theia/wiki/Coding-Guidelines#uri-over-path we should rather replace `PluginModel.packagePath` with `PluginModel.packageUri` if we want to use it on the frontend and backend. cc @benoitf  

path.ts code is correctly aligned with [vscode-uri](https://github.com/microsoft/vscode-uri/blob/master/src/index.ts#L1084) logic and cannot be changed

Comment (owlJaeger) [reply packages/plugin-ext/src/main/browser/plugin-icon-theme-service.ts]: Ok, I'll have a look. That was my 2nd thought to change what's coming in and no try to force it into the right shape. Will update this PR today

Comment (akosyakov) [reply packages/plugin-ext/src/main/browser/plugin-icon-theme-service.ts]: We could add `packageUri` next `packagePath` for now and deprecate the latter without changing other clients for it in this PR. Another PR can clean up replacement of `packagePath` by `packageUri`.

Comment (owlJaeger) [reply packages/plugin-ext/src/main/browser/plugin-icon-theme-service.ts]: At the moment, I've set it up that the old `packageUri` that's now called `packageUriFromPath` has priority over the new `packageUri` if it is defined. If not, the new `packageUri` steps in and everything should work. The checks are still necessary for me though, because it wouldn't work without the front `/` and the `\` being converted to `/` The changes are only made on windows though and shouldn't affect any other operating system.

File: packages/plugin-ext/src/main/browser/plugin-icon-theme-service.ts [hunk 6]
```diff
@@ -540,14 +546,26 @@ export class PluginIconThemeService implements LabelProviderContribution {
     register(contribution: IconThemeContribution, plugin: DeployedPlugin): Disposable {
         const pluginId = getPluginId(plugin.metadata.model);
         const packagePath = plugin.metadata.model.packagePath;
+        const packageUri = (() => {
+            if (isWindows) {
```
Comment (akosyakov) [inline packages/plugin-ext/src/main/browser/plugin-icon-theme-service.ts:550]: that's not what I meant, we should construct URIs on backend. I meant to deprecate `PluginModel.packagePath` with `PluginModel.packageUri`. In this class we can remove `packagePath` and replace it with `packageUri`. I don't think it is going to break anyone.

The issue is that in browser case `isWindows` return for client OS, not backend OS. So if your client is linux, but server runs on windows It won't work.

Comment (owlJaeger) [reply packages/plugin-ext/src/main/browser/plugin-icon-theme-service.ts:550]: Oh. I see. Then I'll update it again to construct the Uri on the backend and pass it on.

Comment (owlJaeger) [reply packages/plugin-ext/src/main/browser/plugin-icon-theme-service.ts]: Should be done

**Comment by owlJaeger** [comment]
Do these changes meet your expectations @akosyakov?

**Comment by akosyakov** [comment]
@owlJaeger it looks very promising, i'm going to test it.

**Comment by owlJaeger** [comment]
Shall I update the Changelog for that? I see there are more deprecation warnings under the `breaking changes` section.

**Comment by akosyakov** [comment]
> Shall I update the Changelog for that? I see there are more deprecation warnings under the breaking changes section.

Yes, you can mention it. Don't delete code yet, only deprecate.

File: packages/plugin-ext/src/main/browser/plugin-icon-theme-service.ts [hunk 7]
```diff
@@ -329,8 +330,10 @@ export class PluginIconTheme extends PluginIconThemeDefinition implements IconTh
             return undefined;
         }
         const iconUri = this.locationUri.resolve(iconPath);
-        const relativePath = this.packageUri.path.relative(iconUri.path.normalize());
-        return relativePath && `url('hostedPlugin/${this.pluginId}/${encodeURIComponent(relativePath.normalize().toString())}')`;
+        const relativePath = this.packageRootUri.path.relative(iconUri.path.normalize());
+        return relativePath && `url('${new Endpoint({
+            path: `hostedPlugin/${this.pluginId}/${encodeURIComponent(relativePath.normalize().toString())}`
+        }).getRestUrl().toString()}')`;
```
Comment (kittaakos) [inline packages/plugin-ext/src/main/browser/plugin-icon-theme-service.ts:336]: The last `toString()` should be `toStirng(true)`, shouldn't it? Otherwise, whitespace will be `%20` in your relative path.

Comment (akosyakov) [reply packages/plugin-ext/src/main/browser/plugin-icon-theme-service.ts:336]: or we should get rid of `encodeURIComponent`, since we go through vscode uri encoding?

**Comment by akosyakov** [comment]
@kittaakos Would you be able to check whether on Windows you have icons for electron?

File: packages/plugin-ext/src/main/browser/plugin-icon-theme-service.ts (hunk 7 above)
Comment (kittaakos) [reply packages/plugin-ext/src/main/browser/plugin-icon-theme-service.ts:336]: No, I was wrong. Ignore this 👆

File: packages/plugin-ext/src/common/plugin-protocol.ts [hunk 8]
```diff
@@ -442,6 +442,7 @@ export interface PluginModel {
         version: string;
     };
     entryPoint: PluginEntryPoint;
+    packageUri: string;
     packagePath: string;
```
Comment (akosyakov) [inline packages/plugin-ext/src/common/plugin-protocol.ts:446]: @owlJaeger please see here how API should be deprecated https://github.com/eclipse-theia/theia/blob/master/doc/api-management.md#deprecation

In this case you need only `@deprecated` tag, no `@since` or `@experimental`. You can read the whole document to learn why.

**Comment by owlJaeger** [comment]
Updated to include the `deprecation warning` and rebased to include newest commits from theia/master.

File: packages/plugin-ext/src/common/plugin-protocol.ts (hunk 8 above)
Comment (owlJaeger) [reply packages/plugin-ext/src/common/plugin-protocol.ts:450]: Done

File: packages/plugin-ext/src/common/plugin-protocol.ts [hunk 9]
```diff
@@ -442,6 +442,11 @@ export interface PluginModel {
         version: string;
     };
     entryPoint: PluginEntryPoint;
+    packageUri: string;
+/**
```
Comment (akosyakov) [inline packages/plugin-ext/src/common/plugin-protocol.ts:446]: please apply our formatting :) `Format Document` action in Theia or VS Code

Comment (owlJaeger) [reply packages/plugin-ext/src/common/plugin-protocol.ts:446]: Oops. You're right. Done

**Comment by akosyakov** [comment]
merging tomorrow if no objections?
